This is a practice for a scRNA analysis.

I used scRNA fastq data of this paper: Piehl, N., van Olst, L., Ramakrishnan, A., Teregulova, V., Simonton, B., Zhang, Z., ... & Gate, D. (2022). Cerebrospinal fluid immune dysregulation during healthy brain aging and cognitive impairment. Cell, 185(26), 5028-5039.

GEO database link is: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE200164

download_cellranger.sh file is a sbatch file. I used sratoolkit to download fastq files from GEO, and cellranger to count the number of transcriptome, used GRCh38 as reference.


seruat.r fils is the r scirpt file used to run seruat function. 
I used mito ratio, number of gene/cell, number of cells, cell cycle distribution as criteria to do QC.  
We also run SCT normalization before doing dimensionality reduction analysis. 
We then runed PCA/UMAP for the clusters generated by KNN. 
